embassy-stm32/src/ospi/mod.rs

```diff
@@ -0,0 +1,1076 @@
+        Self {
+            iwidth: OspiWidth::NONE,
+            instruction: None,
+            isize: AddressSize::_8Bit,
+            idtr: false,
+
+            adwidth: OspiWidth::NONE,
+            address: None,
+            adsize: AddressSize::_8Bit,
+            addtr: false,
+
+            abwidth: OspiWidth::NONE,
+            alternate_bytes: None,
+            absize: AddressSize::_8Bit,
+            abdtr: false,
+
+            dwidth: OspiWidth::NONE,
+            data_len: None,
+            ddtr: false,
+
+            dummy: DummyCycles::_0,
+        }
+    }
+}
+
+/// Error used for Octospi implementation
+#[derive(Debug)]
+pub enum OspiError {
+    /// Peripheral configuration is invalid
+    InvalidConfiguration,
+    /// Operation configuration is invalid
+    InvalidCommand,
+}
+
+/// MultiSpi interface trait
+pub trait MultiSpiBus<Word: Copy + 'static = u8>: embedded_hal_1::spi::ErrorType {
+    /// Transaction configuration for specific multispi implementation
+    type Config;
+
+    /// Command function used for a configuration operation, when no user data is
+    /// supplied to or read from the target device.
+    async fn command(&mut self, config: Self::Config) -> Result<(), Self::Error>;
+
+    /// Read function used to read data from the target device following the supplied transaction
+    /// configuration.
+    async fn read(&mut self, data: &mut [Word], config: Self::Config) -> Result<(), Self::Error>;
+
+    /// Write function used to send data to the target device following the supplied transaction
+    /// configuration.
+    async fn write(&mut self, data: &[Word], config: Self::Config) -> Result<(), Self::Error>;
+}
+
+impl<T: MultiSpiBus<Word> + ?Sized, Word: Copy + 'static> MultiSpiBus<Word> for &mut T {
+    type Config = T::Config;
+    #[inline]
+    async fn command(&mut self, config: Self::Config) -> Result<(), Self::Error> {
+        T::command(self, config).await
+    }
+
+    async fn read(&mut self, data: &mut [Word], config: Self::Config) -> Result<(), Self::Error> {
+        T::read(self, data, config).await
+    }
+
+    async fn write(&mut self, data: &[Word], config: Self::Config) -> Result<(), Self::Error> {
+        T::write(self, data, config).await
+    }
+}
+
+/// OSPI driver.
+pub struct Ospi<'d, T: Instance, Dma> {
+    _peri: PeripheralRef<'d, T>,
+    sck: Option<PeripheralRef<'d, AnyPin>>,
+    d0: Option<PeripheralRef<'d, AnyPin>>,
+    d1: Option<PeripheralRef<'d, AnyPin>>,
+    d2: Option<PeripheralRef<'d, AnyPin>>,
+    d3: Option<PeripheralRef<'d, AnyPin>>,
+    d4: Option<PeripheralRef<'d, AnyPin>>,
+    d5: Option<PeripheralRef<'d, AnyPin>>,
+    d6: Option<PeripheralRef<'d, AnyPin>>,
+    d7: Option<PeripheralRef<'d, AnyPin>>,
+    nss: Option<PeripheralRef<'d, AnyPin>>,
+    dqs: Option<PeripheralRef<'d, AnyPin>>,
+    dma: PeripheralRef<'d, Dma>,
+    config: Config,
+    width: OspiWidth,
+}
+
+impl embedded_hal_1::spi::Error for OspiError {
```

Comment:
please remove these impls too, since this is not an embedded-hal compatible spi anyway